Add input/output to ov ExecNetwork

ilyachur opened the pull request:
### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

ilya-lavrenov commented on inference-engine/tests/functional/plugin/shared/include/behavior/ov_exec_network.hpp:
```diff
@@ -0,0 +1,32 @@
+    ASSERT_EQ(function->inputs().size(), 1);
+    ASSERT_EQ(function->inputs().size(), execNet.inputs().size());
+    ASSERT_NO_THROW(execNet.input());
+    ASSERT_EQ(function->input().get_tensor().get_names(), execNet.input().get_tensor().get_names());
```
can we also compare element types? shapes?

ilyachur replied:
Yes it just a template. First of all I need to fix the behavior 😄

ilya-lavrenov commented on inference-engine/src/inference_engine/src/cpp/ie_executable_network.cpp:
```diff
@@ -136,10 +136,42 @@ std::shared_ptr<const Function> ExecutableNetwork::get_runtime_function() const
     OV_EXEC_NET_CALL_STATEMENT(return _impl->GetExecGraphInfo()->get_results());
 }
+std::vector<ov::Output<const ov::Node>> ExecutableNetwork::outputs() const {
+    const std::shared_ptr<const Function>& func = _impl->GetExecGraphInfo();
```
I believe we cannot use `GetExecGraphInfo` because it does not contain information about tensor names.

Plugins currently does not support them in their execution graph

ilya-lavrenov commented on inference-engine/tests/functional/plugin/shared/include/behavior/ov_exec_network.hpp:
```diff
@@ -5,35 +5,136 @@
+    ASSERT_NE(function->output(1).get_node(), function->output("relu").get_node());
+    ASSERT_EQ(function->output(1).get_node(), function->output("concat").get_node());
+    ASSERT_NE(function->output(0).get_node(), function->output("concat").get_node());
+}
```
Can we finalize with it?

ilyachur replied:
All tests were added

ilya-lavrenov commented on inference-engine/src/inference_engine/src/cnn_network_ngraph_impl.cpp:
```diff
@@ -115,12 +116,17 @@ CNNNetworkNGraphImpl::CNNNetworkNGraphImpl(const std::shared_ptr<Function>& nGra
         info->setInputData(inData);
 
-        if (!_new_api) {
+        if (!_new_api || version == 10) {
```
why do we need it?
For IR v10 and new API we have https://github.com/openvinotoolkit/openvino/pull/7853 which inserts precision conversions as a part of the graph. In that PR I also remove IR version from rt_info as we agreed

ilya-lavrenov commented on inference-engine/src/inference_engine/src/cpp/ie_executable_network.cpp:
```diff
@@ -133,12 +133,82 @@ std::shared_ptr<const Function> ExecutableNetwork::get_runtime_function() const
-    OV_EXEC_NET_CALL_STATEMENT(return _impl->GetExecGraphInfo()->get_parameters());
+std::vector<ov::Output<const ov::Node>> ExecutableNetwork::inputs() const {
+    // clang-format off
+    OV_EXEC_NET_CALL_STATEMENT(
```
we can use `OV_EXEC_NET_CALL_STATEMENT({ xxx })` and clang will provide good formatting

ilya-lavrenov commented on inference-engine/src/inference_engine/src/ie_network_reader.cpp:
```diff
@@ -265,7 +266,7 @@ CNNNetwork details::ReadNetwork(const std::string& modelPath,
     if (inputModel) {
         auto ngFunc = FE->convert(inputModel);
-        return CNNNetwork(ngFunc, exts);
+        return CNNNetwork(ngFunc, exts, newAPI);
```
can we use `CNNNetwork(std::shared_ptr<ICNNNetwork>)` with ngraph impl and don't change public API for `CNNNetwork`?

ilya-lavrenov commented on inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:
```diff
@@ -734,7 +735,7 @@ InferenceEngine::IExecutableNetworkInternal::Ptr Engine::ImportNetwork(std::istr
     CNNNetworkDeserializer deserializer(networkModel,
         [this](const std::string& model, const Blob::CPtr& weights) {
-            return GetCore()->ReadNetwork(model, weights);
+            return CNNNetwork(ov::runtime::Core().read_model(model, weights), {}, true);
```
why do we need this change?

ilya-lavrenov replied:
BTW, new API should accept Tensor instead of Blob :)

ilya-lavrenov commented on inference-engine/src/plugin_api/cpp_interfaces/interface/ie_iexecutable_network_internal.hpp:
```diff
@@ -153,6 +178,8 @@ class INFERENCE_ENGINE_API_CLASS(IExecutableNetworkInternal)
     std::shared_ptr<ov::Function> _runtime_function;  //!< Holds information about network inputs and outputs
```
can we remove this `_runtime_function`?

ilya-lavrenov commented on inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:
```diff
@@ -749,6 +750,7 @@ InferenceEngine::IExecutableNetworkInternal::Ptr Engine::ImportNetwork(std::istr
 
     auto execNetwork = std::make_shared<MKLDNNExecNetwork>(cnnnetwork, conf, extensionManager, weightsSharing);
 
+    SetExeNetworkInfo(execNetwork, cnnnetwork.getFunction());
```
who will call it for other plugins? 
Can we create new inputs / outputs based on existing `GetInputsInfo()`, `GetOutputsInfo()` for import flow?

ilya-lavrenov replied:
let's create task on GNA, KMB, MYRIAD plugins

ilya-lavrenov commented on inference-engine/src/inference_engine/src/cnn_network_ngraph_impl.cpp:
```diff
@@ -274,6 +288,9 @@ StatusCode CNNNetworkNGraphImpl::addOutput(const std::string& layerName,
                 }
+                // Change tensor names for outputs for cases with old API
+                if (!_new_api)
+                    result->output(0).get_tensor().set_names({result->get_friendly_name()});
```
can we add test for this case?

ilya-lavrenov commented on inference-engine/tests/functional/plugin/shared/include/behavior/ov_exec_network.hpp:
```diff
@@ -51,21 +166,217 @@ TEST_P(OVExecNetwork, precisionsAsInOriginalIR) {
+    ASSERT_THROW(importedExecNet.output("relu").get_node(), ov::Exception);
+    ASSERT_THROW(importedExecNet.output("concat").get_node(), ov::Exception);
+    ASSERT_NO_THROW(importedExecNet.output("relu_op").get_node());
+    ASSERT_NO_THROW(importedExecNet.output("concat_op").get_node());
```
can we also check that precision are imported correctly? i.e. with CNNNetwork's conversion is applied
but need to change original precisions to f16

ilya-lavrenov commented on inference-engine/src/mkldnn_plugin/mkldnn_serialize.cpp:
```diff
@@ -117,8 +117,14 @@ void CNNNetworkSerializer::operator << (const CNNNetwork & network) {
+    // Serialize to old representation in case of old API
+    if (network.newAPI()) {
+        ngraph::pass::StreamSerialize serializer(_ostream, getCustomOpSets(), serializeInputsAndOutputs);
+        serializer.run_on_function(std::const_pointer_cast<ngraph::Function>(network.getFunction()));
```
can we have `GetCore()->isNewAPI()` method instead of patching `CNNNetwork`?

ilya-lavrenov commented on inference-engine/tests/functional/plugin/cpu/shared_tests_instances/skip_tests_config.cpp:
```diff
@@ -91,6 +93,9 @@ std::vector<std::string> disabledTestPatterns() {
         R"(.*smoke_(Auto|Multi)_BehaviorTests/OVExecNetwork.*type=i8.*)",
+        R"(.*smoke_(Auto|Multi)_BehaviorTests.*OVExecNetwork.*importExportedNetwork.*)",
+        R"(.*smoke_(Auto|Multi)_BehaviorTests.*OVExecNetwork.*importExportedIENetwork.*)",
+        R"(.*smoke_(Auto|Multi)_BehaviorTests.*OVExecNetwork.*readFromV10IR.*)",
```
can we also enable tests with HETERO?

ilya-lavrenov commented on inference-engine/tests/functional/plugin/shared/include/behavior/ov_exec_network.hpp:
```diff
@@ -51,21 +166,217 @@ TEST_P(OVExecNetwork, precisionsAsInOriginalIR) {
+    ASSERT_EQ(importedExecNet.inputs().size(), 1);
+    ASSERT_EQ(importedExecNet.outputs().size(), 1);
+    ASSERT_NO_THROW(importedExecNet.input("in1"));
+    ASSERT_NO_THROW(importedExecNet.output("round"));
```
can we also check that precision are imported correctly? i.e. with CNNNetwork's conversion is applied
but need to change original precisions to f16

ilya-lavrenov commented on inference-engine/src/inference_engine/src/ie_network_reader.cpp:
```diff
@@ -9,13 +9,37 @@
+            auto dataName = ngraph::op::util::create_ie_output_name(result->input_value(0));
+            result->output(0).get_tensor().set_names({dataName});
+        }
+    }
```
Maybe the problem here is that we serialize function from IR v10 as IR v11 in new API?

ilyachur commented:
@ilya-lavrenov @nosovmik Please take a look.
I fixed the order of parameters/results after the pre-processing
And removed an erase of version from the function
